### ISSUE TITLE
Enforcing stability of event rates in time

### ISSUE THREAD

**Issue by vavrato** [guillermo-navas-palencia/optbinning#216]
Hi,

in business, it may be important to have stability of event rates in time.
For instance, if event rate for bin_1 in the January sample is smaller than event rate for bin_2 in the January sample, we want the same inequality for all the months.

Without understanding all the detail, I think that this could be implemented at least when monotonicity is enforced (because in this case we know the relative positions of event rates a priori. Namely, it would (probably?) mean adding $t$ constraints ( $t$ being the number of observed timestamps) of the type from the inequality below (6) here https://arxiv.org/pdf/2001.08025.pdf

**Comment by guillermo-navas-palencia** [comment]
Correct: 1 year per scenario.

OptimalBinning will return a higher IV in most cases unless the data and target distribution per scenario are identical. The advantages of the scenario-based are:

- same bins for all scenarios.
- imposed constraints are satisfied for all scenarios.

Could you share your data/code to understand better what you are aiming to do?

**Comment by vavrato** [comment]
I'll be damned, trying to reproduce my example, but there might have been an error of some sort on my side yesterday.

What I got yesterday was two different binnings from OptimalBinning and SBOptimalBinning, both with two bins, both feasible solutions (identical target distribution per scenario), but different split. Hence I thought that SBOptimalBinning does not maximize IV.

Now I realized that the error has been made when I was preparing the data for the SBOB where I accidentally used a subset of the samples.

So it does seem now that SBOB does maximize IV while keeping the bins event rates order across scenarios. Which is great. This is a very useful feature, maybe you should advertise it more; this functionality is not very apparent from the description through scenarios.

**Comment by vavrato** [comment]
Just to make sure: SBOptimalBinning does not support categorical features, right? If so, I have a feature request :-)

**Comment by guillermo-navas-palencia** [comment]
You are right, categorical features are not supported. Could you please open a new GitHub issue with this request, and close this one if the problem is solved?

I agree a better or additional tutorial is needed to show the possibilities of the SBOptimalBinning.

**Comment by vavrato** [comment]
I am reopening this thread because I have more questions if you don't mind.

- I do not understand how SBOB works. Trying it out on testing data, I get vastly different binning with monotonic_trend='descending' than None, while None gives a feasible solution (and a better one) already. Please see the attached picture (the plot for event rates of bins in time/scenarios), the data are also attached.  monotonic_trend = 'peak'.
- I thought SBOB would do what you can see in the picture -- no 'crossing' of event rates in time/scenarios. Also while maximizing total IV. Is that how it is supposed to work?
- Setting min_bin_size = 0.05 gives a different binning than the default value which should be the same value ( = min_prebin_size = 0.05). Am I missing something?
![Binnings](https://user-images.githubusercontent.com/25643293/203832385-c122039e-9e61-4116-ac2e-851f5ac497b5.png)
[example.csv](https://github.com/guillermo-navas-palencia/optbinning/files/10086278/example.csv)

**Comment by vavrato** [comment]
For your convenience:

```
example = pd.read_csv('example.csv')
sbo = SBOptimalBinning(max_n_bins = 5, monotonic_trend = 'descending') # or some other setting
grouped = example.groupby('month')
Xs = [i['feature'] for _,i in grouped]
ys = [i['target'] for _,i in grouped]
sbo.fit(Xs,ys)
```

**Comment by guillermo-navas-palencia** [comment]
I found the bug. There is a pre-solving algorithm applicable to the general case, OptimalBinning, but it does not work for SBOptimalBinning. After removing it, it returns the same binning as "valley" or "peak". Thanks for testing it!

I will clarify your questions later.

**Comment by guillermo-navas-palencia** [comment]
> I am reopening this thread because I have more questions if you don't mind.
> 
> * I do not understand how SBOB works. Trying it out on testing data, I get vastly different binning with monotonic_trend='descending' than None, while None gives a feasible solution (and a better one) already. Please see the attached picture (the plot for event rates of bins in time/scenarios), the data are also attached.  monotonic_trend = 'peak'.
> * I thought SBOB would do what you can see in the picture -- no 'crossing' of event rates in time/scenarios. Also while maximizing total IV. Is that how it is supposed to work?
> * Setting min_bin_size = 0.05 gives a different binning than the default value which should be the same value ( = min_prebin_size = 0.05). Am I missing something?
>   ![Binnings](https://user-images.githubusercontent.com/25643293/203832385-c122039e-9e61-4116-ac2e-851f5ac497b5.png)
>   [example.csv](https://github.com/guillermo-navas-palencia/optbinning/files/10086278/example.csv)

- Regarding your first comment, as I mentioned earlier, a pre-solving technique was erroneously applied to the SBOptimalBinning, making some feasible solutions infeasible. This error will be corrected in the next release. Note that this error only occurs when monotonic_trend is set to 'ascending', 'descending', 'peak_heuristic' and 'valley_heuristic'.
- The non-crossing constraint is not implemented, and it would require various changes in the current implementation. Having said so, I did implement a similar algorithm in the past, so it is doable. The SBOptimalBinning algorithm generates a binning solution maximizing the total IV (sum of total IV across all scenarios) while satisfying constraints for all scenarios. Taking your example:

1. Global solution
![image](https://user-images.githubusercontent.com/40094942/203872853-9cf42f71-2097-48c8-aa24-eabba677e715.png)
2. Binning for each scenario:
![image](https://user-images.githubusercontent.com/40094942/203872891-af152eae-50cb-4c1b-833f-47d46d4820e8.png) 
![image](https://user-images.githubusercontent.com/40094942/203872908-8d004702-7696-4388-b78b-71c35f58be1b.png)
![image](https://user-images.githubusercontent.com/40094942/203872960-736ddfae-5462-4afc-a643-74601858b6e7.png)
![image](https://user-images.githubusercontent.com/40094942/203873012-2799ff43-b703-403d-8291-8c9ff846ff32.png)
![image](https://user-images.githubusercontent.com/40094942/203873035-9872ba91-bcec-4caa-a9e8-34ec767be5e6.png)

- The constraints are implemented at scenario-level. The 5% is not satisfied with the last scenario:
![image](https://user-images.githubusercontent.com/40094942/203873154-5b067123-346e-4b45-9c76-9242d21c79da.png)
Setting min_bin_size to 0.05 returns:
![image](https://user-images.githubusercontent.com/40094942/203873243-65d7afb5-2cbc-4ee0-837f-94c5bbd3761c.png)

Perhaps, I didn't pay much attention to this algorithm since I thought it would not be used or appreciated. Since there seems to be some interest, I will devote more time to improving documentation and examples (+ adding new features).

**Comment by guillermo-navas-palencia** [comment]
To clarify, by non-crossing I refer to this plot:
![image](https://user-images.githubusercontent.com/40094942/203875747-9a3c4e33-63e7-4e96-9431-d4244ea38bbf.png)

However, your plot will be non-crossing by default:
![image](https://user-images.githubusercontent.com/40094942/203875805-d434182a-a6b8-4000-8e0d-cbc473263ac8.png)

**Comment by vavrato** [comment]
Thank you for your explanation.

The non-crossing property from the last answer is not what I had in mind and it is hard for me now to come with a purpose for such a thing. On the other hand, the same trend/ordering for each scenario applied to time is something very useful.
Please could you check the attached data? (same format as yesterday). Even for peak, valley, convex, concave I get different orderings across the scenarios (months) which I think should be an unfeasible solution?
[example.csv](https://github.com/guillermo-navas-palencia/optbinning/files/10094800/example.csv)

**Comment by guillermo-navas-palencia** [comment]
I checked it. There is no error in the definition of these 4 monotonic trends. Strictly speaking, for >= 2 segments, ascending or descending could also be valley or peak (the change point is the last or first bin). The case convex/concave is similar, by definition convex/concave could also be ascending, descending, peak, or valley.

I assume you were expecting a common change point, but that is not enforced, so the obtained result is feasible.

**Comment by vavrato** [comment]
Thanks for the explanation. I understand I can get the functionality that I want only with the ascending or descending monotonic trend.

I guess enforcing a common change point in peak/valley might result in a non-linear programming problem?

**Comment by guillermo-navas-palencia** [comment]
A common peak/valley can be implemented partially reusing the method for normal binning. The non-linearities (disjoint constraints) are linearized.

I might not have time to implement it for v0.17.2, but I will try to include it in v0.18.0.

**Comment by guillermo-navas-palencia** [comment]
I just released version 0.17.2 with the bugfixes commented on this issue. Please open a new issue for the stability of categorical features. Thanks!
